BUILD: Hide platform configuration probe behind --debug-configure

mattip opened the pull request:
As requested in #14470, this adds a `--debug-configure` option to `runtests.py` and `setup.py build` to hide all the compiler output during the configuration probe to create `config.h` and `_numpyconfig.h`.

This also converts some of the direct `print` to stdout so that it uses the logger instead. The lack of the flag simply raises the log level to `WARN` from `INFO` for the duration of the noisy function.

mattip committed:
BUILD: add --debug-configure option to reduce output, use logger more

mattip committed:
BUILD: only report these headers with --debug-configure

mattip committed:
BUILD: fix runtests.py option, fix previous changeset

mattip commented:
Should I add the `--debug-configure` flag to the CI runs?

rgommers commented on numpy/distutils/command/build.py:
```diff
@@ -16,8 +16,8 @@ class build(old_build):
     user_options = old_build.user_options + [
         ('fcompiler=', None,
          "specify the Fortran compiler type"),
-        ('parallel=', 'j',
```
This change seems unrelated, why does `parallel` need to be removed here and in `finalize_options`?

mattip replied:
It is redundant - the base `build` class has this option (on python 3)

mattip replied:
Specifically it was added in CPython v3.5.0

rgommers replied:
Ah yes. We may have to add it back once we figure out why it's buggy right now, but let's remove it till someone gets annoyed enough to start looking into that.

rgommers commented on numpy/core/setup.py:
```diff
@@ -638,23 +638,6 @@ def generate_api(ext, build_dir):
-    # build_clib cannot handle generate_config_h and generate_numpyconfig_h
-    # (don't ask). Because clib are generated before extensions, we have to
-    # explicitly add an extension which has generate_config_h and
-    # generate_numpyconfig_h as sources *before* adding npymath.
```
what's going on here, did all this become obsolete somehow?

mattip replied:
It should have been removed when we merged `umath` and `multiarray`. It causes the config file generation to be run twice, which I noticed when working on this PR.

rgommers commented:
> Should I add the `--debug-configure` flag to the CI runs?

Probably a good idea. The `SDIST=1` build should use the new default (hide config output), in the other CI builds it's mostly just noise I think. Unless other people have a use for it, I can't remember many such cases ....

rgommers commented:
Thanks for tackling this, will be very nice to have!

mattip commented:
Comparing the logs from another PR's travis run to this one, the lines from

[building library "npysort" sources](https://travis-ci.org/numpy/numpy/jobs/585230589#L3055) to
[building extension "numpy.core._multiarray_tests" sources](https://travis-ci.org/numpy/numpy/jobs/585230589#L4664) now become from

[building library "npysort" sources](https://travis-ci.org/numpy/numpy/jobs/585258939#L3022) to
[building extension "numpy.core._multiarray_tests" sources](https://travis-ci.org/numpy/numpy/jobs/585258939#L3037), removing about 1600 lines.

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -162,7 +165,7 @@ def CCompiler_spawn(self, cmd, display=None):
         cmd = ' '.join(list(cmd))
 
-    forward_bytes_to_stdout(o)
+    log.info(o)
```
I think there's a problem here that `o` has type `bytes`, but log.info expects `str`. Previously we got around this by assuming we could pass GCC's output directly to the terminal without decoding and re-encoding, which saved us having to guess GCC's encoding.

mattip replied:
fixing (make it conditional on `self.verbose`)

mattip committed:
BUILD: mostly restore verbose output on CI (except for pip install)

mattip committed:
BUG: fixes from review

eric-wieser commented on numpy/distutils/ccompiler.py:
```diff
@@ -18,7 +18,7 @@
 from numpy.distutils.compat import get_exception
 from numpy.distutils.exec_command import (
-    filepath_from_subprocess_output, forward_bytes_to_stdout
+    filepath_from_subprocess_output
```
I think you need this import back

mattip replied:
fixed

mattip commented:
failing because of #14526

mattip committed:
MAINT: move the option to build_src, simplifies logic

stefanv commented on runtests.py:
```diff
@@ -67,6 +67,8 @@ def main(argv):
     parser.add_argument("--verbose", "-v", action="count", default=1,
                         help="more verbosity")
+    parser.add_argument("--debug-configure", action="store_true",
+                        help="add -v to build_src to show cconfiguration compiler output")
```
This is pretty vague; can we describe what the actual feature is that is enabled?

mattip replied:
"adds -v to the build_src subcommand which will show compiler output while probing platform features used to create _numpyconfig.h and config.h" seems a bit too much. 

Any ideas what would be more appropriate?

stefanv replied:
I think the problem is that there are several things being addressed here: `-v` is being added to `build_src`, but only for a certain type of build?  Or is it always added?  If always, then the latter part of the sentence is not necessary.  It can simply be something like "Increase verbosity of build_src by adding -v flag".  But the "configure" in `--debug-configure` makes me think it was intended for something more specific.

mattip replied:
another option would be to not add it to runtests.py at all (which would always build without the flag), and in CI explicitly do `setup.py build build_src -v; runtests.py`.

rgommers replied:
> and in CI explicitly do `setup.py build build_src -v; runtests.py`

That doesn't seem desirable at all, we want to run CI in the ways we expect humans to use things, not some arcane invocation. There's no real issue here, so why change anything?

stefanv replied:
Yes, I think just a slightly more descriptive help string would be sufficient.

mattip committed:
DOC, BUILD: make build_src more discoverable

rgommers commented:
CI failure `python3: can't open file 'build': [Errno 2] No such file or directory` is related.

I'm -1 on the interface changes and promoting `build_src` usage, this was good before.

mattip committed:
DOC: improve help string

mattip commented:
rebased and reverted the change. The help string now reads
```
...
optional arguments:
  -h, --help            show this help message and exit
  --verbose, -v         more verbosity
  --debug-configure     add -v to build_src to show compiler configuration
                        output while creating _numpyconfig.h and config.h
  --no-build, -n        do not build the project (use system installed
                        version)
...
```
```
clarifications happily accepted.

rgommers commented:
That help description seems very clear to me, thanks @mattip.

rgommers commented:
In it goes, very nice to have this!

rgommers commented:
Hmm, one of the Windows builds on Azure for the merge commit failed, see https://dev.azure.com/numpy/numpy/_build/results?buildId=5577. 

Related to gh-14464, which may have a bad interaction with this PR. Any thoughts @mattip?

```
================================== FAILURES ===================================
____________________ TestMatmul.test_vector_matrix_values _____________________

self = <numpy.core.tests.test_multiarray.TestMatmul object at 0x14B88330>

    def test_vector_matrix_values(self):
        vec = np.array([1, 2])
        mat1 = np.array([[1, 2], [3, 4]])
        mat2 = np.stack([mat1]*2, axis=0)
        tgt1 = np.array([7, 10])
        tgt2 = np.stack([tgt1]*2, axis=0)
        for dt in self.types[1:]:
            v = vec.astype(dt)
            m1 = mat1.astype(dt)
            m2 = mat2.astype(dt)
>           res = self.matmul(v, m1)
E           RuntimeWarning: invalid value encountered in matmul

dt         = 'F'
m1         = array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]], dtype=complex64)
m2         = array([[[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]],

       [[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]]], dtype=complex64)
mat1       = array([[1, 2],
       [3, 4]])
mat2       = array([[[1, 2],
        [3, 4]],

       [[1, 2],
        [3, 4]]])
res        = array([[ 7., 10.],
       [ 7., 10.]], dtype=float64)
self       = <numpy.core.tests.test_multiarray.TestMatmul object at 0x14B88330>
tgt1       = array([ 7, 10])
tgt2       = array([[ 7, 10],
       [ 7, 10]])
v          = array([1.+0.j, 2.+0.j], dtype=complex64)
vec        = array([1, 2])

C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\site-packages\numpy\core\tests\test_multiarray.py:6031: RuntimeWarning
____________________ TestMatmul.test_matrix_vector_values _____________________

self = <numpy.core.tests.test_multiarray.TestMatmul object at 0x14C75330>

    def test_matrix_vector_values(self):
        vec = np.array([1, 2])
        mat1 = np.array([[1, 2], [3, 4]])
        mat2 = np.stack([mat1]*2, axis=0)
        tgt1 = np.array([5, 11])
        tgt2 = np.stack([tgt1]*2, axis=0)
        for dt in self.types[1:]:
            v = vec.astype(dt)
            m1 = mat1.astype(dt)
            m2 = mat2.astype(dt)
>           res = self.matmul(m1, v)
E           RuntimeWarning: invalid value encountered in matmul

dt         = 'F'
m1         = array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]], dtype=complex64)
m2         = array([[[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]],

       [[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]]], dtype=complex64)
mat1       = array([[1, 2],
       [3, 4]])
mat2       = array([[[1, 2],
        [3, 4]],

       [[1, 2],
        [3, 4]]])
res        = array([[ 5., 11.],
       [ 5., 11.]], dtype=float64)
self       = <numpy.core.tests.test_multiarray.TestMatmul object at 0x14C75330>
tgt1       = array([ 5, 11])
tgt2       = array([[ 5, 11],
       [ 5, 11]])
v          = array([1.+0.j, 2.+0.j], dtype=complex64)
vec        = array([1, 2])

C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\site-packages\numpy\core\tests\test_multiarray.py:6058: RuntimeWarning
________________ TestMatmulOperator.test_vector_matrix_values _________________

self = <numpy.core.tests.test_multiarray.TestMatmulOperator object at 0x14C545B0>

    def test_vector_matrix_values(self):
        vec = np.array([1, 2])
        mat1 = np.array([[1, 2], [3, 4]])
        mat2 = np.stack([mat1]*2, axis=0)
        tgt1 = np.array([7, 10])
        tgt2 = np.stack([tgt1]*2, axis=0)
        for dt in self.types[1:]:
            v = vec.astype(dt)
            m1 = mat1.astype(dt)
            m2 = mat2.astype(dt)
>           res = self.matmul(v, m1)
E           RuntimeWarning: invalid value encountered in matmul

dt         = 'F'
m1         = array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]], dtype=complex64)
m2         = array([[[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]],

       [[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]]], dtype=complex64)
mat1       = array([[1, 2],
       [3, 4]])
mat2       = array([[[1, 2],
        [3, 4]],

       [[1, 2],
        [3, 4]]])
res        = array([[ 7., 10.],
       [ 7., 10.]], dtype=float64)
self       = <numpy.core.tests.test_multiarray.TestMatmulOperator object at 0x14C545B0>
tgt1       = array([ 7, 10])
tgt2       = array([[ 7, 10],
       [ 7, 10]])
v          = array([1.+0.j, 2.+0.j], dtype=complex64)
vec        = array([1, 2])

C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\site-packages\numpy\core\tests\test_multiarray.py:6031: RuntimeWarning
________________ TestMatmulOperator.test_matrix_vector_values _________________

self = <numpy.core.tests.test_multiarray.TestMatmulOperator object at 0x14B88470>

    def test_matrix_vector_values(self):
        vec = np.array([1, 2])
        mat1 = np.array([[1, 2], [3, 4]])
        mat2 = np.stack([mat1]*2, axis=0)
        tgt1 = np.array([5, 11])
        tgt2 = np.stack([tgt1]*2, axis=0)
        for dt in self.types[1:]:
            v = vec.astype(dt)
            m1 = mat1.astype(dt)
            m2 = mat2.astype(dt)
>           res = self.matmul(m1, v)
E           RuntimeWarning: invalid value encountered in matmul

dt         = 'F'
m1         = array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]], dtype=complex64)
m2         = array([[[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]],

       [[1.+0.j, 2.+0.j],
        [3.+0.j, 4.+0.j]]], dtype=complex64)
mat1       = array([[1, 2],
       [3, 4]])
mat2       = array([[[1, 2],
        [3, 4]],

       [[1, 2],
        [3, 4]]])
res        = array([[ 5., 11.],
       [ 5., 11.]], dtype=float64)
self       = <numpy.core.tests.test_multiarray.TestMatmulOperator object at 0x14B88470>
tgt1       = array([ 5, 11])
tgt2       = array([[ 5, 11],
       [ 5, 11]])
v          = array([1.+0.j, 2.+0.j], dtype=complex64)
vec        = array([1, 2])
```

mattip commented:
That is the infamous heisenbug that occurs 1/1000.

charris commented:
This broke the appveyor wheel builds. See https://ci.appveyor.com/project/matthew-brett/numpy-wheels/builds/27571654/job/v5tm7ljl6htcfa0a. Probably the changes in `setup.py` put things out of order.
